lib/listen/record.rb

```diff
@@ -75,31 +57,42 @@ def build
 
     private
 
-    def _auto_hash
-      Hash.new { |h, k| h[k] = Hash.new }
+    # TODO: refactor/refactor out
+    def add_dir(dir, rel_path)
+      rel_path = '.' if [nil, '', '.'].include? rel_path
+      dirname, basename = Pathname(rel_path).split.map(&:to_s)
+      basename = '.' if [nil, '', '.'].include? basename
+      root = (@paths[dir.to_s] ||= {})
+      if [nil, '', '.'].include?(dirname)
+        entries = (root['.'] || {})
+        entries.merge!(basename => {}) if basename != '.'
+        root['.'] = entries
+      else
+        root[rel_path] ||= {}
+      end
```

Comment:
This fails to add the directory name to the parent folder's entry list.
For example, if we have a simple directory containing one folder, `path`, and within that, `folder`, and within that a file called `README`, the resulting tree would be:

```
'.' => {'path' => {}}
'path' => {}
'path/folder' => {'README' => {<file-data>}}
```

It should be:

```
'.' => {'path' => {}}
'path' => {'folder' => {}}
'path/folder' => {'README' => {<file-data>}}
```

Otherwise when `Directory::scan` is processing deletion for `path`, it will never recurse `folder` and not actually raise any deletion events.

This seems to work, and I'll look at making a test unless you're able to find it easier?

```
    def add_dir(dir, rel_path)
      rel_path = '.' if [nil, '', '.'].include? rel_path
      dirname, basename = Pathname(rel_path).split.map(&:to_s)
      basename = '.' if [nil, '', '.'].include? basename
      root = (@paths[dir.to_s] ||= {})
      dirname = '.' if [nil, '', '.'].include?(dirname)
      entries = (root[dirname] || {})
      entries.merge!(basename => {}) if basename != '.'
      root[dirname] = entries
    end
```